### PR TITLE
[SYSSETUP] Set registry hostname in WriteComputerSettings

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Based on @Doug-Lyons's patch.
JIRA issue: [CORE-16067](https://jira.reactos.org/browse/CORE-16067)

### CODE REVIEW

File: dll/win32/syssetup/wizard.c [hunk 1]
```diff
@@ -560,6 +562,32 @@ WriteComputerSettings(WCHAR * ComputerName, HWND hwndDlg)
...
 
+    /* Now we need to set the Hostname */
+    lError = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                           L"SYSTEM\\CurrentControlSet\\Services\\Tcpip\\Parameters",
+                           0,
+                           KEY_SET_VALUE,
+                           &hKey);
+    if (lError != ERROR_SUCCESS)
+    {
+        DPRINT1("RegOpenKeyExW(\"HKLM\\System\\CurrentControlSet\\Services\\Tcpip\\Parameters\") failed!\n");
```
Comment (learn-more) [inline dll/win32/syssetup/wizard.c]: It would be useful to include the error code here

File: dll/win32/syssetup/wizard.c [hunk 2]
```diff
@@ -560,6 +562,32 @@ WriteComputerSettings(WCHAR * ComputerName, HWND hwndDlg)
...
+
+    lError = RegSetValueEx(hKey,
+                           L"Hostname",
+                           0,
+                           REG_SZ,
+                           (LPBYTE)ComputerName,
+                           (wcslen(ComputerName) + 1) * sizeof(WCHAR));
+    if (lError != ERROR_SUCCESS)
+    {
+        DPRINT1("RegSetValueEx(\"Hostname\") failed!\n");
```
Comment (learn-more) [inline dll/win32/syssetup/wizard.c]: Same here.

**Comment by Doug-Lyons** [comment]
@katahiromz  Thank you. This looks great.

File: dll/win32/syssetup/wizard.c [hunk 3]
```diff
@@ -560,6 +562,32 @@ WriteComputerSettings(WCHAR * ComputerName, HWND hwndDlg)
...
+    lError = RegSetValueEx(hKey,
+                           L"Hostname",
+                           0,
+                           REG_SZ,
+                           (LPBYTE)ComputerName,
+                           (wcslen(ComputerName) + 1) * sizeof(WCHAR));
+    if (lError != ERROR_SUCCESS)
+    {
+        DPRINT1("RegSetValueEx(\"Hostname\") failed (%08lX)\n", lError);
+        return FALSE;
```
Comment (SergeGautherie) [inline dll/win32/syssetup/wizard.c]: Missing `RegCloseKey(hKey);`.

File: dll/win32/syssetup/wizard.c [hunk 4]
```diff
@@ -586,28 +614,31 @@ WriteDefaultLogonData(LPWSTR Domain)
...
                            L"DefaultDomainName",
                            0,
                            REG_SZ,
                            (LPBYTE)Domain,
-                           (wcslen(Domain)+ 1) * sizeof(WCHAR));
+                           (wcslen(Domain) + 1) * sizeof(WCHAR));
     if (lError != ERROR_SUCCESS)
     {
-        DPRINT1("RegSetValueEx(\"DefaultDomainName\") failed!\n");
+        DPRINT1("RegSetValueEx(\"DefaultDomainName\") failed (%08lX)\n", lError);
```
Comment (SergeGautherie) [inline dll/win32/syssetup/wizard.c]: The current function returns `TRUE`, even when there are failures.
Either fix it or add a comment about that behavior?

File: dll/win32/syssetup/wizard.c (hunk 3 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: OK.

File: dll/win32/syssetup/wizard.c (hunk 4 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: Fixed.

Comment (Doug-Lyons) [reply dll/win32/syssetup/wizard.c]: @SergeGautherie  and @katahiromz This function returning FALSE if there is an error with the Registry changes will cause the setup screen to not advance to the next step and there is nothing that the user can do to fix either of these last two error, so I believe that returning TRUE and allowing advancing to the next setup step is the correct behavior here. For an example of what happens just change a character is the registry path to break this and then try building from a clean start and you will see what I mean. Basically it is impossible to install ReactOS under these circumstances. Thanks.

**Comment by Doug-Lyons** [comment]
@SergeGautherie and @katahiromz Please see my comment regarding having this function return FALSE under Serge's comments above. Thanks.

**Comment by SergeGautherie** [comment]
> return FALSE

I don't know what the wanted behavior(s) is: are/should (all) these errors blocking or ignorable?
If ignorable, I am simply suggesting explicit code comments.

Some in-consistency clues on existing code, for this PR or a follow-up:
- SetComputerNameW: MessageBoxW + FALSE
- SetComputerNameExW: ignored
- RegOpenKeyExW: FALSE

Doug, aren't existing FALSE cases already blocking install in exactly the same way?
As in no regression, only further improvement to do?
Note that hopefully these failures should "never" happen in real life. As in Minor.

**Comment by Doug-Lyons** [comment]
@SergeGautherie At present the only way for the function WriteComputerSettings to return FALSE is for the SetComputerNameW function to fail and when this happens, you receive a popup messagebox showing that this has happened and you are returned to the Setup Window where you can correct this. For any other failures that the user cannot correct we should return TRUE in my opinion. Thanks.

File: dll/win32/syssetup/wizard.c [hunk 5]
```diff
@@ -560,7 +563,33 @@ WriteComputerSettings(WCHAR * ComputerName, HWND hwndDlg)
...
+    lError = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                           L"SYSTEM\\CurrentControlSet\\Services\\Tcpip\\Parameters",
+                           0,
+                           KEY_SET_VALUE,
+                           &hKey);
+    if (lError != ERROR_SUCCESS)
+    {
+        DPRINT1("RegOpenKeyExW for Tcpip\\Parameters failed (%08lX)\n", lError);
+        return FALSE;
+    }
```
Comment (Doug-Lyons) [inline dll/win32/syssetup/wizard.c]: Please return TRUE here.

File: dll/win32/syssetup/wizard.c [hunk 6]
```diff
@@ -560,7 +563,33 @@ WriteComputerSettings(WCHAR * ComputerName, HWND hwndDlg)
...
+    lError = RegSetValueEx(hKey,
+                           L"Hostname",
+                           0,
+                           REG_SZ,
+                           (LPBYTE)ComputerName,
+                           (wcslen(ComputerName) + 1) * sizeof(WCHAR));
+    if (lError != ERROR_SUCCESS)
+    {
+        DPRINT1("RegSetValueEx(\"Hostname\") failed (%08lX)\n", lError);
+        ret = FALSE;
```
Comment (Doug-Lyons) [inline dll/win32/syssetup/wizard.c]: Please return TRUE here.

File: dll/win32/syssetup/wizard.c [hunk 7]
```diff
@@ -560,7 +563,33 @@ WriteComputerSettings(WCHAR * ComputerName, HWND hwndDlg)
...
+    /* Now we need to set the Hostname */
+    lError = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                           L"SYSTEM\\CurrentControlSet\\Services\\Tcpip\\Parameters",
+                           0,
+                           KEY_SET_VALUE,
+                           &hKey);
+    if (lError != ERROR_SUCCESS)
+    {
+        DPRINT1("RegOpenKeyExW for Tcpip\\Parameters failed (%08lX)\n", lError);
+        return FALSE;
```
Comment (Doug-Lyons) [inline dll/win32/syssetup/wizard.c]: Please return TRUE here.

File: dll/win32/syssetup/wizard.c [hunk 8]
```diff
@@ -560,7 +563,33 @@ WriteComputerSettings(WCHAR * ComputerName, HWND hwndDlg)
...
+                           (wcslen(ComputerName) + 1) * sizeof(WCHAR));
+    if (lError != ERROR_SUCCESS)
+    {
+        DPRINT1("RegSetValueEx(\"Hostname\") failed (%08lX)\n", lError);
+        ret = FALSE;
+    }
+
+    RegCloseKey(hKey);
+
+    return ret;
```
Comment (Doug-Lyons) [inline dll/win32/syssetup/wizard.c]: Please return TRUE here.

File: dll/win32/syssetup/wizard.c (hunk 6 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: OK.

File: dll/win32/syssetup/wizard.c (hunk 7 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: OK.

File: dll/win32/syssetup/wizard.c (hunk 8 above)
Comment (katahiromz) [reply dll/win32/syssetup/wizard.c]: OK.

**Comment by SergeGautherie** [comment]
> the only way for the function WriteComputerSettings to return FALSE

"Yes", but what about WriteDefaultLogonData() case?

> For any other failures that the user cannot correct we should return TRUE in my opinion.

That feels wrong to me as is: please, do explain why it is correct to simply ignore these errors, as in no user feedback and "broken" install.

**Comment by Doug-Lyons** [comment]
@SergeGautherie , the WriteDefaultLogonData() function is not something that I have touched in my changes. Therefore, I have not studied it in detail and cannot offer any comments either way for now. I suggest that if you want to improve this function, then please submit another PR for your changes.
As far as "simply ignore these errors", they are not ignored, but captured in the debug log. It is true that we have not brought them to the users attention, but the user can only provide reports of problems based on any popup messages. If there are errors because of these changes they will probably show up later and when a developer reviews the debug log these errors will be seen then. Stopping the user from installing ReactOS just because of obvious programmer errors just discourages testers. Thanks.

**Comment by Doug-Lyons** [comment]
@katahiromz Thank you very much for your help with this. Doug Lyons

**Pull request merged**